### PR TITLE
python3Packages.fakeredis: add patch to support redis<=4.1.0

### LINKED ISSUE #156815: `python39Packages.fakeredis` is currently broken
### Describe the bug
As of 4d922730369d1c468cc4ef5b2fc30fd4200530e0, the build of `python39Packages.fakeredis` is broken. This is currently breaking the build of other packages including 
* `python3Packages.wandb`
* `python3Packages.sentry-sdk`
* `cura`
* `moz-phab`
* `octoprint`
* `gns3-server`
* `gns3-gui`
* `tribler`
* `python3Packages.django-rq`
* `alerta-server`
* `moodle-dl`
and certainly many others...

~I suspect this was triggered by https://github.com/NixOS/nixpkgs/pull/140482, #131015, or one of the redis upgrades.~ I believe this was triggered by #152550.

### Steps To Reproduce
1. Checkout nixpkgs at 4d922730369d1c468cc4ef5b2fc30fd4200530e0
2. `nix-build -A python39Packages.fakeredis`
3. Observe failure
```
...
ERROR: Could not find a version that satisfies the requirement redis<4.1.0 (from fakeredis) (from versions: none)
ERROR: No matching distribution found for redis<4.1.0
```

### Expected behavior
The build to succeed.

### Screenshots
n/a

### Additional context
n/a

### Notify maintainers

@fabaff @SuperSandro2000 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.91, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.335376.610d4ea2750, nix-ld"`
 - channels(skainswo): `"home-manager, nixpkgs-unstable-22.05pre343295.adf7f03d3bf"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by fabaff**
###### Motivation for this change
Fixes #156815

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 156851` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>22 packages failed to build:</summary>
  <ul>
    <li>alerta-server</li>
    <li>cura</li>
    <li>gns3-gui</li>
    <li>gns3-server</li>
    <li>home-assistant-component-tests.sentry</li>
    <li>moodle-dl</li>
    <li>moz-phab</li>
    <li>python310Packages.django-rq</li>
    <li>python310Packages.sentry-sdk</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.wandb</li>
    <li>python39Packages.django-rq</li>
    <li>python39Packages.sentry-sdk</li>
    <li>python39Packages.spacy</li>
    <li>python39Packages.spacy-loggers</li>
    <li>python39Packages.spacy-transformers</li>
    <li>python39Packages.textacy</li>
    <li>python39Packages.wandb</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>octoprint</li>
    <li>python310Packages.chalice</li>
    <li>python310Packages.fakeredis</li>
    <li>python39Packages.chalice</li>
    <li>python39Packages.fakeredis</li>
  </ul>
</details>

File: pkgs/development/python-modules/fakeredis/default.nix [hunk 1]
```diff
@@ -42,6 +45,15 @@ buildPythonPackage rec {
     pytestCheckHook
   ];
 
+  patches = [
+    # Support for redis >= 4.1.0, https://github.com/jamesls/fakeredis/pull/324
+    (fetchpatch {
+      name = "support-redis-4.1.0.patch";
+      url = "https://github.com/jamesls/fakeredis/commit/8ef8dc6dacc9baf571d66a25ffbf0fadd7c70f78.patch";
```
Comment (samuela) [inline pkgs/development/python-modules/fakeredis/default.nix:52]: I believe that this patch actually adds support for `redis<=4.1.0` not `redis>=4.1.0`. Would you mind updating the commit message?

File: pkgs/development/python-modules/fakeredis/default.nix [hunk 2]
```diff
@@ -19,11 +20,13 @@ buildPythonPackage rec {
...
+  disabled = pythonOlder "3.7";
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "sha256-yb0S5DAzbL0+GJ+uDpHrmZl7k+dtv91u1n+jUtxoTHE=";
+  src = fetchFromGitHub {
+    owner = "jamesls";
+    repo = pname;
+    rev = version;
+    hash = "sha256-P6PUg9SY0Qshlvj+iV1xdrzVLJ9JXUV4cGHUynKO3m0=";
```
Comment (samuela) [inline pkgs/development/python-modules/fakeredis/default.nix:29]: Thanks for switching to the source build!

File: pkgs/development/python-modules/fakeredis/default.nix [hunk 3]
```diff
@@ -19,11 +20,13 @@ buildPythonPackage rec {
   version = "1.7.0";
   format = "setuptools";
 
-  disabled = pythonOlder "3.5";
+  disabled = pythonOlder "3.7";
```
Comment (samuela) [inline pkgs/development/python-modules/fakeredis/default.nix:23]: Is there a reason that this was changed?

**Review by samuela — CHANGES_REQUESTED** [review]
Thanks for putting this together! I'd def like to get it merged ASAP

File: pkgs/development/python-modules/fakeredis/default.nix (hunk 1 above)
Comment (fabaff) [reply pkgs/development/python-modules/fakeredis/default.nix:52]: Done

File: pkgs/development/python-modules/fakeredis/default.nix (hunk 2 above)
Comment (fabaff) [reply pkgs/development/python-modules/fakeredis/default.nix:29]: Was need to use the patch.

File: pkgs/development/python-modules/fakeredis/default.nix (hunk 3 above)
Comment (fabaff) [reply pkgs/development/python-modules/fakeredis/default.nix:23]: `unstable` doesn't support Python 3.6 or 3.5 anymore out-of-the-box thus I can only check if the package build locally on Python 3.7 and above. Also, 3.6 is EOL.

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 156851` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>22 packages failed to build:</summary>
  <ul>
    <li>alerta-server</li>
    <li>cura</li>
    <li>gns3-gui</li>
    <li>gns3-server</li>
    <li>home-assistant-component-tests.sentry</li>
    <li>moodle-dl</li>
    <li>moz-phab</li>
    <li>python310Packages.django-rq</li>
    <li>python310Packages.sentry-sdk</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.wandb</li>
    <li>python39Packages.django-rq</li>
    <li>python39Packages.sentry-sdk</li>
    <li>python39Packages.spacy</li>
    <li>python39Packages.spacy-loggers</li>
    <li>python39Packages.spacy-transformers</li>
    <li>python39Packages.textacy</li>
    <li>python39Packages.wandb</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>octoprint</li>
    <li>python310Packages.chalice</li>
    <li>python310Packages.fakeredis</li>
    <li>python39Packages.chalice</li>
    <li>python39Packages.fakeredis</li>
  </ul>
</details>

**Review by samuela — APPROVED** [review]
LGTM I'll merge tomorrow unless anyone objects.

**Pull request merged**